File: OpenConvo/Assets/Plugins/OpenConvo/OCQuests.js
#pragma strict

import System.Collections.Generic;

public class OCQuests {
	public class Giver {
		public var name : String;
		public var portrait : Texture2D;

		function QuestGiver ( name : String, portrait : Texture2D ) {
			this.name = name;
			this.portrait = portrait;
		}
	}

	public class Objective {
		public var description : String;
		public var progress : int = 0;
		public var goal : int = 1;

		function Objective ( description : String, goal : int ) {
			if ( goal < 1 ) {
				goal = 1;
			}
			
			this.description = description;
			this.goal = goal;
		}	

		public function get completed () : boolean {
			if ( progress > goal ) {
				progress = goal;
			}

			return progress == goal;
		}

		public function set completed ( value : boolean ) {
			progress = value ? goal : 0;
		}
	}
	
	public class Quest {
		public var title : String;
		public var description : String;
		public var objectives : Objective[] = new Objective[0];
		public var xp : int;
		public var giver : Giver;
		public var image : Texture2D;
		public var side : boolean;

		function Quest () {
		}

		function Quest ( title : String, description : String, objectives : Objective[] ) {
			Quest ( title, description, objectives, 0, false, null, null );
		}
		
		function Quest ( title : String, description : String, objectives : Objective[], xp : int ) {
			Quest ( title, description, objectives, xp, false, null, null );
		}
		
		function Quest ( title : String, description : String, objectives : Objective[], xp : int, side : boolean ) {
			Quest ( title, description, objectives, xp, side, null, null );
		}

		function Quest ( title : String, description : String, objectives : Objective[], xp : int, side : boolean, giver : Giver ) {
			Quest ( title, description, objectives, xp, side, giver, null );
		}

		function Quest ( title : String, description : String, objectives : Objective[], xp : int, side : boolean, giver : Giver, image : Texture2D ) {
			this.title = title;
			this.description = description;
			this.objectives = objectives;
			this.xp = xp;
			this.side = side;
			this.giver = giver;
			this.image = image;
		}

		public function get completed () : boolean {
			var completedObjectives : int = 0;

			for ( var i : int = 0; i < objectives.Length; i++ ) {
				if ( objectives[i].completed ) {
					completedObjectives++;
				}
			}

			if ( objectives.Length > 0 && completedObjectives == objectives.Length ) {
				return true;
		
			} else {
				return false;
			}
		}

		public function set completed ( value : boolean ) {
			for ( var i : int = 0; i < objectives.Length; i++ ) {
				objectives[i].completed = value;
			}
		}

		public function ProgressObjective ( description : String, amount : int ) {
			for ( var i : int = 0; i < objectives.Length; i++ ) {
				if ( description == objectives[i].description ) {
					ProgressObjective ( i, amount );
					break;
				}
			}
		}

		public function ProgressObjective ( i : int, amount : int ) {
			if ( i >= 0 && i < objectives.Length ) {
				objectives[i].progress += amount;

				if ( objectives[i].progress >= objectives[i].goal ) {
					CompleteObjective ( i );
				}
			}			
		}

		public function CompleteObjective ( i : int ) {
			if ( i >= 0 && i < objectives.Length ) {
				objectives[i].completed = true;

				var manager : OCManager = OCManager.GetInstance();

				if ( manager ) {
					manager.eventHandler.OnObjectiveCompleted ( this, i );
				
				} else {
					Debug.LogError ( "OCManager not in scene!" );

				}
			}
		}

		public function GetObjectiveStrings ( maxLength : int ) : String [] {
			var strings : String [] = new String [ objectives.Length ];

			for ( var i : int = 0; i < objectives.Length; i++ ) {
				strings[i] = objectives[i].description;

				if ( strings[i].Length > maxLength ) {
					strings[i] = strings[i].Substring ( 0, maxLength ) + "...";
				}
			}

			return strings;
		}
	}

	public var userQuests : Quest[] = new Quest[0];
	public var potentialQuests : Quest[] = new Quest[0];
	public var activeQuest : int = -1;
	
	public function SetActiveQuest ( quest : Quest ) {
		for ( var i : int = 0; i < userQuests.Length; i++ ) {
			if ( userQuests[i] == quest ) {
				activeQuest = i;
				return;
			}
		}

		activeQuest = -1;
	}

	public function GetActiveQuest () : Quest {
		if ( activeQuest > 0 && activeQuest < userQuests.Length ) {
			return userQuests[activeQuest];
		}

		return null;
	}

	public function AddUserQuest ( quest : Quest ) {
		var tmp : List.< Quest > = new List.< Quest > ( userQuests );

		tmp.Add ( quest );

		userQuests = tmp.ToArray ();
	}

	public function RemoveUserQuest ( quest : Quest ) {
		var tmp : List.< Quest > = new List.< Quest > ( userQuests );

		tmp.Remove ( quest );

		userQuests = tmp.ToArray ();
	}
	
	public function AddPotentialQuest ( quest : Quest ) {
		var tmp : List.< Quest > = new List.< Quest > ( potentialQuests );

		tmp.Add ( quest );

		potentialQuests = tmp.ToArray ();
	}

	public function RemovePotentialQuest ( quest : Quest ) {
		var tmp : List.< Quest > = new List.< Quest > ( potentialQuests );

		tmp.Remove ( quest );

		potentialQuests = tmp.ToArray ();
	}

	public function GetUserQuest ( title : String ) : Quest {
		for ( var i : int = 0; i < userQuests.Length; i++ ) {
			if ( userQuests[i].title == title ) {
				return userQuests[i];
			}
		}

		return null;
	}
	
	public function GetPotentialQuest ( title : String ) : Quest {
		for ( var i : int = 0; i < potentialQuests.Length; i++ ) {
			if ( potentialQuests[i].title == title ) {
				return potentialQuests[i];
			}
		}

		return null;
	}

	public function IsQuestCompleted ( title : String ) : boolean {
		var quest : Quest = GetUserQuest ( title );

		if ( quest ) {
			return quest.completed;

		} else {
			return false;
				
		}
	}
}

File: OpenConvo/Assets/Plugins/OpenConvo/OCQuests.cs
using UnityEngine;
using System.Collections;

using System.Collections.Generic;

[System.Serializable]
public class OCQuests {
	[System.Serializable]
	public class Giver {
		public string name;
		public Texture2D portrait;

		public Giver ( string name, Texture2D portrait ) {
			this.name = name;
			this.portrait = portrait;
		}
	}

	[System.Serializable]
	public class Objective {
		public string description;
		public int progress = 0;
		public int goal = 1;

		public Objective ( string description, int goal ) {
			if ( goal < 1 ) {
				goal = 1;
			}
			
			this.description = description;
			this.goal = goal;
		}	

		public bool completed {
			get {
				if ( progress > goal ) {
					progress = goal;
				}

				return progress == goal;
			}
			
			set {
				progress = (bool)value ? goal : 0;
			}
		}
	}
	
	[System.Serializable]
	public class Quest {
		public string title;
		public string description;
		public Objective[] objectives = new Objective[0];
		public int xp;
		public Giver giver;
		public Texture2D image;
		public bool side;

		public Quest () {
		
		}

		public Quest ( string title, string description, Objective[] objectives, int xp = 0, bool side = false, Giver giver = null, Texture2D image = null ) {
			this.title = title;
			this.description = description;
			this.objectives = objectives;
			this.xp = xp;
			this.side = side;
			this.giver = giver;
			this.image = image;
		}

		public bool completed {
			get {
				int completedObjectives = 0;

				for ( int i = 0; i < objectives.Length; i++ ) {
					if ( objectives[i].completed ) {
						completedObjectives++;
					}
				}

				return objectives.Length > 0 && completedObjectives == objectives.Length;
			}

			set {
				for ( int i = 0; i < objectives.Length; i++ ) {
					objectives[i].completed = (bool)value;
				}
			}
		}

		public void ProgressObjective ( string description, int amount ) {
			for ( int i = 0; i < objectives.Length; i++ ) {
				if ( description == objectives[i].description ) {
					ProgressObjective ( i, amount );
					break;
				}
			}
		}

		public void ProgressObjective ( int i, int amount ) {
			if ( i >= 0 && i < objectives.Length ) {
				objectives[i].progress += amount;

				if ( objectives[i].progress >= objectives[i].goal ) {
					CompleteObjective ( i );
				}
			}			
		}

		public void CompleteObjective ( int i ) {
			if ( i >= 0 && i < objectives.Length ) {
				objectives[i].completed = true;

				OCManager manager = OCManager.GetInstance();

				if ( manager ) {
					manager.eventHandler.OnObjectiveCompleted ( this, i );
				
				} else {
					Debug.LogError ( "OCManager not in scene!" );

				}
			}
		}

		public string [] GetObjectiveStrings ( int maxLength ) {
			string[] strings = new string [ objectives.Length ];

			for ( int i = 0; i < objectives.Length; i++ ) {
				strings[i] = objectives[i].description;

				if ( strings[i].Length > maxLength ) {
					strings[i] = strings[i].Substring ( 0, maxLength ) + "...";
				}
			}

			return strings;
		}
	}

	public Quest[] userQuests = new Quest[0];
	public Quest[] potentialQuests = new Quest[0];
	public int activeQuest = -1;
	
	public string[] GetQuestNames () {
		string[] strings = new string [ potentialQuests.Length ];
		
		for ( int i = 0; i < strings.Length; i++ ) {
			strings[i] = potentialQuests[i].title;
		}

		return strings;
	}
	
	public string[] GetObjectiveNames ( string quest ) {
		List< string > strings = new List< string > ();
		
		for ( int i = 0; i < potentialQuests.Length; i++ ) {
			if ( quest == potentialQuests[i].title ) {
				for ( int o = 0; o < potentialQuests[i].objectives.Length; o++ ) {
					strings.Add ( potentialQuests[i].objectives[o].description );
				}
				break;
			}
		}

		return strings.ToArray ();
	}

	public int GetIndex ( string title ) {
		for ( int i = 0; i < potentialQuests.Length; i++ ) {
			if ( potentialQuests[i].title == title ) {
				return i;
			}
		}
		
		return 0;
	}
	
	public void SetActiveQuest ( Quest quest ) {
		for ( int i = 0; i < userQuests.Length; i++ ) {
			if ( userQuests[i] == quest ) {
				activeQuest = i;
				return;
			}
		}

		activeQuest = -1;
	}

	public Quest GetActiveQuest () {
		if ( activeQuest >= 0 && activeQuest < userQuests.Length ) {
			return userQuests[activeQuest];
		}

		return null;
	}

	public void AddUserQuest ( Quest quest ) {
		List< Quest > tmp = new List< Quest > ( userQuests );

		activeQuest = tmp.Count;

		tmp.Add ( quest );

		userQuests = tmp.ToArray ();
	}

	public void RemoveUserQuest ( Quest quest ) {
		List< Quest > tmp = new List< Quest > ( userQuests );

		tmp.Remove ( quest );

		userQuests = tmp.ToArray ();
	}
	
	public void AddPotentialQuest ( Quest quest ) {
		List< Quest > tmp = new List< Quest > ( potentialQuests );

		tmp.Add ( quest );

		potentialQuests = tmp.ToArray ();
	}

	public void RemovePotentialQuest ( Quest quest ) {
		List< Quest > tmp = new List< Quest > ( potentialQuests );

		tmp.Remove ( quest );

		potentialQuests = tmp.ToArray ();
	}

	public Quest GetUserQuest ( string title ) {
		for ( int i = 0; i < userQuests.Length; i++ ) {
			if ( userQuests[i].title == title ) {
				return userQuests[i];
			}
		}

		return null;
	}
	
	public Quest GetPotentialQuest ( string title ) {
		for ( int i = 0; i < potentialQuests.Length; i++ ) {
			if ( potentialQuests[i].title == title ) {
				return potentialQuests[i];
			}
		}

		return null;
	}

	public bool IsQuestCompleted ( string title ) {
		Quest quest = GetUserQuest ( title );

		if ( quest != null ) {
			return quest.completed;

		} else {
			return false;
				
		}
	}
}
